File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        unordered_map<char, char> matching = {
            {')', '('},
            {']', '['},
            {'}', '{'}
        };
        // int l=0,r=s.size()-1;
        // while(l<=r){
        //     if(matching[s[r]] == s[l]){
        //         l++;
        //         r--;
        //     }
        //     else{
        //         return false;
        //     }
        // }
        // return true;
        // ABOVE FAILS ON TESTCASE-> I/P: s="()[]{}"


        // Intuition and Understanding:
        // Use a stack to simulate nested structure. Push open brackets, pop and match with closing brackets.
        // If at the end the stack is empty and all matches are correct, return true.

        for(char ch:s){

            if(matching.count(ch)){
                // if closing bracket
                    // match with the top of stack
                if(st.empty() || st.top() != matching[ch])return false;
                st.pop();
            }
            else{
                //if opening bracket-> push to stack
                st.push(ch);
            }
        }
        return st.empty();
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    # my solution:
    # def isValid(self, s: str) -> bool:
    #     if (len(s))%2 != 0:return False
    #     stk = deque()
    #     closed = [')', ']', '}']
    #     mp = {
    #         '(': ')', 
    #         '[': ']',
    #         '{': '}',
    #     }
    #     for ch in s:
    #         if ch in closed:
    #             if len(stk)>0:
    #                 x = stk[-1]
    #                 stk.pop()
    #                 if mp[x] != ch:
    #                     return False
    #             else:
    #                 #no open brackets in stk and closing bracket came
    #                 return False
    #         else:
    #             stk.append(ch)
    #     if len(stk) != 0: 
    #         return False #open brackets remaining
    #     return True

    # cleaner:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {')': '(', ']': '[', '}': '{'}
        for char in s:
            if char in mapping:
                if not stack or stack.pop() != mapping[char]:
                    return False
            else:
                stack.append(char)
        return not stack